### PR TITLE
XRI support for the new input system (Action base interactions + Simulator) TRNG-1349

### PR DESCRIPTION
**Pull request by Gusinuhe**
### Description

This PR adds support for the new Input System.

### Type of change

- [X] New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Replace the old `XR Rig` according to the input system configured in the project:

Use `[XR_Setup_Action_Based]` when testing the `old input system`.
Use `[XR_Setup_Device_Based]` when testing the `new input system`.

> When setting the project is set to `both`, both prefabs should work fine. 

For testing the simulator, set the project to use either the `new input system` or `both`, go to `Package Manager` and search for the `XR Interaction Toolkit` package, and import the `XR Simulator`:

image

> More [info and details](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@0.10/manual/samples.html)

A new folder should be added to the project, search for the `XR Device Simulator` prefab, add it to the scene next to the `[XR_Setup_Action_Based]`, make sure the project is configured to be either in `new input system` or `both` and that no VR configuration is enabled.

For more details on the mapping of the simulator take a closer look at the [XR Device Simulator Controls](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@0.10/manual/samples.html#input-actions-asset-1) file or [this video](https://youtu.be/d4bTpkvBwrs), sorry no official documentation´ 😟.

### Todo

- [X] XRI support for the old input system.
- [X] XRI support for the new input system.
- [X] Support for the simulator.
- [X] Add UI controller for the XR Rig action-based.
- [X] Remove warnings for the new input system.
- [X] Add support for the standalone course controller when using the new input system.
- [X] Make the Advance course controller (template UI overlay) work when using the new input system.

### CODE REVIEW

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+    /// Use this class to map input actions to each controller state (mode)
+    /// and set up the transitions between controller states (modes).
+    /// </summary>
+    [AddComponentMenu("XR/Action Based Controller Manager")]
+    [DefaultExecutionOrder(ControllerManagerUpdateOrder)]
+    public class ActionBasedControllerManager : MonoBehaviour
+    {
+        private const int ControllerManagerUpdateOrder = 10;
+
+        public enum StateID
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:22]: xmldoc

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+
+        /// <summary>
+        /// Find and configure the components on the teleport controller.
+        /// </summary>
+        /// <param name="enable"> Set it true to enable the teleport controller, false to disable it. </param>
+        private void SetTeleportController(bool enable)
+        {
+            FindTeleportControllerComponents();
+
+            if (teleportLineVisual != null)
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:548]: teleportLineVisual?.enabled = enable

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+
+        /// <summary>
+        /// Find and configure the components on the UI controller.
+        /// </summary>
+        /// <param name="enable"> Set it true to enable the UI controller, false to disable it. </param>
+        private void SetUIController(bool enable)
+        {
+            FindUIControllerComponents();
+
+            if (uiLineVisual != null)
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:572]: Same

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs [hunk 4]
```diff
@@ -0,0 +1,787 @@
...
+            }
+            
+            // Transition from Select state to Teleport state when the user triggers the "Teleport Mode Activate" action but not the "Cancel Teleport" action
+            InputAction teleportModeAction = GetInputAction(teleportModeActivate);
+            InputAction cancelTeleportModeAction = GetInputAction(teleportModeCancel);
+            InputAction uiModeAction = GetInputAction(uiModeActivate);
+
+            bool triggerTeleportMode = teleportModeAction != null && teleportModeAction.triggered;
+            bool cancelTeleport = cancelTeleportModeAction != null && cancelTeleportModeAction.triggered;
+            bool triggerUIMode = uiModeAction != null && uiModeAction.triggered;
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:686]: confused me a bit, could be isUIModeTriggered to be clear etc.

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs [hunk 5]
```diff
@@ -0,0 +1,787 @@
...
+        /// <summary>
+        /// Updated every frame to handle the transition to m_SelectState state.
+        /// </summary>
+        private void OnUpdateTeleportState()
+        {
+            // Transition from Teleport state to Select state when we release the Teleport trigger or cancel Teleport mode
+            InputAction teleportModeAction = GetInputAction(teleportModeActivate);
+            InputAction cancelTeleportModeAction = GetInputAction(teleportModeCancel);
+
+            bool cancelTeleport = cancelTeleportModeAction != null && cancelTeleportModeAction.triggered;
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:717]: isCancelTeleportTriggered

File: Runtime/Interaction/Action-based/LocomotionSchemeManager.cs [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+                    throw new InvalidEnumArgumentException(nameof(scheme), (int)scheme, typeof(MoveSchemeType));
+            }
+        }
+
+        private void SetTurnStyle(TurnStyleType style)
+        {
+            switch (style)
+            {
+                case TurnStyleType.Snap:
+                    if (continuousTurnProvider != null)
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/LocomotionSchemeManager.cs:373]: Feels like the switch consists of null checks, one liners would improve the overview.

File: Runtime/Interaction/Action-based/LocomotionSchemeManager.cs [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+                }
+            }
+        }
+
+        private void ClearBindingMasks()
+        {
+            SetBindingMasks(string.Empty);
+        }
+
+        private InputControlScheme? FindControlScheme(string controlSchemeName, InputActionReference action)
```
Comment (SimonTheSourcerer) [inline Runtime/Interaction/Action-based/LocomotionSchemeManager.cs:497]: those 2 classes feel like code duplication

**Review by SimonTheSourcerer — CHANGES_REQUESTED** [review]
Is there any docu required for this?

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs (hunk 2 above)
Comment (Gusinuhe) [reply Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:548]: `The left-hand side of an assignment must be a variable, property or indexer` 🤷‍♂️
image

File: Runtime/Interaction/Action-based/ActionBasedControllerManager.cs (hunk 3 above)
Comment (Gusinuhe) [reply Runtime/Interaction/Action-based/ActionBasedControllerManager.cs:572]: Same as above

**Comment by Gusinuhe** [comment]
All should now work when tested along with https://github.com/Innoactive/Basic-UI-Component/pull/9

**Review by SimonTheSourcerer — DISMISSED** [review]

**Review by SimonTheSourcerer — APPROVED** [review]

**Pull request merged**